### ISSUE TITLE
Default memory hint is too aggressive during pruning

### ISSUE THREAD

**Issue by yorickdowne** [NethermindEth/nethermind#5548]
**Description**

With Nethermind set to Hybrid mode pruning and default values for prune cache and memory hint, on a 32 GiB RAM machine with Lighthouse, a Full prune will OOM Nethermind.

Setting memory hint to 1 GiB resolves this.

**Expected behavior**

With memory hint at default (null), Nethermind uses sufficiently conservative settings to be able to prune without OOM.

**Desktop**

 - Operating System: Linux
 - Version: 1.17.3+671d60ad
 - Installation Method: docker
 - Consensus Client: Lighthouse 4.0.1

**Comment by yorickdowne** [comment]
After some testing, recommend changing default memory hint for `mainnet.cfg` to `1024000000` instead of `2048000000`. That way, prune won't OOM. Memory hint for `mainnet_archive.cfg` could stay as-is.

**Comment by yorickdowne** [comment]
I was asked to test this with prune mode full instead of Hybrid and memory hint 2048000000. That OOMs.

![image](https://user-images.githubusercontent.com/71337066/231447490-2f0e7499-7195-4eac-ab20-1ddbdad573ac.png)

**Comment by asdacap** [comment]
Seems like `VerifyTrie` (same code) can OOM relatively easily with batched trie visitor. Does not seems to be affected by memory hint. Reducing parallelism or turning on direct reads also works, which either means some kind of bug in rocksdb/OS if it runs too fast or the os cache got too big if it run too fast.

**Comment by asdacap** [comment]
Limiting max open files also works.

**Comment by asdacap** [comment]
@yorickdowne how many core/thread do you have?

**Comment by yorickdowne** [comment]
6 cores, 12 threads, max parallelism set to half cores: 3

**Comment by asdacap** [comment]
Hmm.. what is you SSD model. And to double check, you are using linux right?

**Comment by yorickdowne** [comment]
Yep Linux, here a Debian 11 but I also have some Ubuntu 22.04. The SSD is a Samsung PM9A3.

**Comment by asdacap** [comment]
Does not happen again after a resync... This issue is highly inconsistent.

**Comment by yorickdowne** [comment]
Interesting. I have a previously-pruned node that just OOM'd on me during normal operation, not during a prune. Logs show nothing unusual. It gets a payload, it looks happy, and then OOM and restart:

```
eth-main-execution-1  | 2023-04-18 13:55:34.8200|Peers | with known best block: 50 | all: 50 | 
eth-main-execution-1  | 2023-04-18 13:55:39.5315|Received a new payload: 17073991 (0xd5c542c22df6c76c17b95ef0a7a3230cc0df732462a50572e97b80a2804ceb0d) 
eth-main-execution-1  | 2023-04-18 13:55:40.1827|Processed   17073991 |      646ms of  14,988ms, mgasps   46.30 total  110.14, tps  467.65 total 1029.29, bps    1.55 total    3.62, recv queue 0, proc queue 0 
eth-main-execution-1  | 2023-04-18 13:55:40.1827|Valid. Result of a new payload: 17073991 (0xd5c542c22df6c76c17b95ef0a7a3230cc0df732462a50572e97b80a2804ceb0d). 
eth-main-execution-1  | Using pruning parameters:
eth-main-execution-1  | --Pruning.FullPruningMaxDegreeOfParallelism=3 --Pruning.Mode=Full --Pruning.FullPruningTrigger=VolumeFreeSpace --Pruning.FullPruningThresholdMb=375810
eth-main-execution-1  | 2023-04-18 13-55-45.1968|Nethermind starting initialization.
eth-main-execution-1  | 2023-04-18 13-55-45.2578|Loading embedded plugins
```

**Comment by asdacap** [comment]
Before it OOM, did it restart after prune?

**Comment by yorickdowne** [comment]
Yes, this is configured to restart after prune. ` --Pruning.FullPruningCompletionBehavior AlwaysShutdown`

**Comment by asdacap** [comment]
@yorickdowne does it still happen with 1.18.0?

**Comment by yorickdowne** [comment]
This is resolved by 1.18.0
